app/src/main/java/org/simple/clinic/registration/phone/RegistrationPhoneScreenController.kt

```diff
@@ -52,15 +55,24 @@ class RegistrationPhoneScreenController @Inject constructor(
     //
     // The registration phone screen already opens if the user has not requested an otp
     // yet. This just ensures the local stored user is cleared to restart the flow.
-    val clearLocallyStoredUser = events
+    //
+    // One exception where we don't clear the user is if the user's logged in status
+    // is UNAUTHORIZED, because this means that this screen has been opened to
+    // ask the user to sign in again, and this is handled elsewhere.
+    return events
         .ofType<RegistrationPhoneScreenCreated>()
+        .flatMapSingle { userSession.loggedInUser().firstOrError() }
+        .filterAndUnwrapJust()
+        .filter { loggedInUser -> loggedInUser.loggedInStatus != User.LoggedInStatus.UNAUTHORIZED }
         .flatMap {
           userSession.clearLoggedInUser().andThen(Observable.empty<UiChange>())
```

Comment:
I agree with you. However, refactoring is a little tricky because `Flow` currently requires us to provide a key when initialising it and this is the key that is provided in most cases. This also needs to be provided in a blocking fashion and is invoked on the main thread, so we need to avoid database calls on this, so this clearing of data was being done right here.

I have added a maintenance story to come back to this at some point. https://www.pivotaltracker.com/story/show/166607823

Reply:
I actually meant that we could think of clearing data based on the current state of the user and not based on the UI changes like `RegistrationPhoneScreenCreated`.
Let's discuss this in-person.